types/src/units.rs

```diff
@@ -0,0 +1,114 @@
+use async_trait::async_trait;
+use codec::Codec;
+
+/// An abstraction for reading saved data that can be Encoded and Decoded
+#[async_trait]
+pub trait Reader<T: Codec + Send + Sync>: Send + Sync {
+    /// Returns Some(T) containing next element to read.
+    /// Return None if there is no elements left to read.
+    fn next(&mut self) -> Option<T>;
+}
+
+/// An abstraction for saving data that can be Encoded and Decoded
+#[async_trait]
+pub trait Backup<T: Codec + Send + Sync>: Send + Sync {
+    /// Saves data `T`. Upon ending needs to guarantee that `data` is saved.
+    async fn save(&self, data: T);
+}
+
+/// A type to generate a structure for saving data that needs to be backed up.
+/// Example implementation:
+/// ```rust
+/// struct BackupImpl {
+///     tx: UnboundedSender<Vec<u8>>,
+/// }
+///
+/// #[async_trait]
+/// impl<T: Codec + Send + Sized + Sync + 'static> Backup<T> for BackupImpl {
+///     async fn save(&self, data: T) {
+///         if let Err(e) = self.tx.unbounded_send(data.encode()) {
+///             ...
+///         }
+///     }
+/// }
+///
+/// struct IntoBackupImpl {
+///     tx: UnboundedSender<Vec<u8>>,
+/// }
+///
+/// impl IntoBackup for IntoBackupImpl {
+///     type Into = BackupImpl;
+///
+///     fn into_backup<T: Codec + Send + Sized + Sync + 'static>(self) -> Self::Into {
+///         Backup { tx: self.tx }
+///     }
+/// }
+///
+/// impl IntoBackupImpl {
+///     fn new(tx: UnboundedSender<Vec<u8>>) -> IntoBackupImpl {
+///         IntoBackupImpl { tx }
+///     }
+/// }
+/// ```
+pub trait IntoBackup {
```

Comment:
Do you even need this trait? Couldn't you `impl<T: ..., B: Backup<T>> Into<B> for IntoBackupImpl`?

Same for the reader ofc.